forms/Form.php

```diff
@@ -1181,31 +1189,66 @@ public function setMessage($message, $type, $escapeHtml = true) {
 	 *                            user supplied data in the message.
 	 */
 	public function sessionMessage($message, $type, $escapeHtml = true) {
-		Session::set(
-			"FormInfo.{$this->FormName()}.formError.message", 
-			$escapeHtml ? Convert::raw2xml($message) : $message
-		);
-		Session::set("FormInfo.{$this->FormName()}.formError.type", $type);
+		// Benign message
+		if($type == "good") {
+			$this->getSessionValidationResult()->addMessage($message, $type, null, $escapeHtml);
+
+		// Bad message causing a validation error
+		} else {
+			$this->getSessionValidationResult()->addError($message, $type, null, $escapeHtml
+			);
+		}
 	}
 
-	public static function messageForForm($formName, $message, $type, $escapeHtml = true) {
-		Session::set(
-			"FormInfo.{$formName}.formError.message", 
-			$escapeHtml ? Convert::raw2xml($message) : $message
-		);
-		Session::set("FormInfo.{$formName}.formError.type", $type);
+	/**
+	 * @deprecated 3.1
+	 */
+	public static function messageForForm($formName, $message, $type) {
+		Deprecation::notice('3.1', 'Create an instance of the form you wish to attach a message to.');
+	}
+
+	/**
+	 * Returns the ValidationResult stored in the session.
+	 * You can use this to modify messages without throwing a ValidationException.
+	 * If a ValidationResult doesn't yet exist, a new one will be created
+	 *
+	 * @return ValidationResult The ValidationResult object stored in the session
+	 */
+	public function getSessionValidationResult() {
+		$result = Session::get("FormInfo.{$this->FormName()}.result");
```

Comment:
I know that we can store serialised objects in the session, but couldn't this be a lot more data than we need? (Given ValidationResult extends Object, and all of that data is potentially being serialised along with it).

I'd be happy if ValidationResult implemented Serializable so that I knew my session wasn't being filled up with garbage. :)